File: bt_service/loader/template.py
"""Tree template loader — scans directories for .json tree files."""

import json
import os

from bt_service.core.registry import Registry
from bt_service.factory.registry import FactoryRegistry
from bt_service.factory.deserialize import deserialize_tree


def load_tree_dir(dir_path: str, reg: FactoryRegistry, tree_reg: Registry) -> list[dict]:
    """Scan dir_path for .json files, deserialize each, register in tree_reg.

    Returns list of {name, source} for inspection tools.
    """
    sources = []
    if not os.path.isdir(dir_path):
        return sources

    for entry in sorted(os.listdir(dir_path)):
        if not entry.endswith(".json"):
            continue
        filepath = os.path.join(dir_path, entry)
        with open(filepath, "r", encoding="utf-8") as f:
            data = f.read()

        node, _ = deserialize_tree(data, reg)
        name = entry[:-5]  # strip .json
        tree_reg.register(name, node)
        sources.append({"name": name, "source": data})
    return sources


def validate_tree_json(data: str, reg: FactoryRegistry) -> tuple[bool, str]:
    """Validate JSON tree without registering. Returns (valid, error_msg)."""
    try:
        json.loads(data)  # ensure valid JSON
    except json.JSONDecodeError as e:
        return False, str(e)
    try:
        deserialize_tree(data, reg)
        return True, ""
    except (ValueError, KeyError, TypeError) as e:
        return False, str(e)

File: pkg/bt/template.go
package bt

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// TreeSource holds a deserialized tree along with its original JSON source.
type TreeSource struct {
	Name   string
	Source json.RawMessage
}

// LoadTreeDir scans a directory for .json tree files, deserializes each,
// and registers them in the tree Registry. Returns the list of loaded sources
// for inspection tools.
//
// Each .json file must be a valid TreeFile format (name + tree).
func LoadTreeDir(dir string, reg *FactoryRegistry, treeReg *Registry) ([]TreeSource, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	var sources []TreeSource
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		if filepath.Ext(entry.Name()) != ".json" {
			continue
		}

		data, err := os.ReadFile(filepath.Join(dir, entry.Name()))
		if err != nil {
			return nil, fmt.Errorf("read %s: %w", entry.Name(), err)
		}

		// Validate JSON before trying to deserialize
		if !json.Valid(data) {
			return nil, fmt.Errorf("invalid JSON in %s", entry.Name())
		}

		node, _, err := DeserializeTree(data, reg)
		if err != nil {
			return nil, fmt.Errorf("deserialize %s: %w", entry.Name(), err)
		}

		// Use filename without extension as tree name
		name := entry.Name()[:len(entry.Name())-5]
		treeReg.Register(name, node)

		sources = append(sources, TreeSource{
			Name:   name,
			Source: data,
		})
	}

	return sources, nil
}

// ValidateTreeJSON parses and builds a tree from JSON bytes without
// registering it. Returns nil on success, or an error describing the issue.
func ValidateTreeJSON(data []byte, reg *FactoryRegistry) error {
	if !json.Valid(data) {
		return fmt.Errorf("invalid JSON")
	}
	_, _, err := DeserializeTree(data, reg)
	return err
}
